packages/compiler/server/serverlib.ts

```diff
@@ -1121,57 +1120,94 @@ export function createServer(host: ServerHost): Server {
     return url ? host.getOpenDocumentByURL(url) : undefined;
   }
 
-  async function readFile(path: string): Promise<ServerSourceFile> {
-    // Try open files sent from client over LSP
-    const document = getOpenDocument(path);
-    if (document) {
-      return {
-        document,
-        ...createSourceFile(document.getText(), path),
-      };
-    }
+  async function fileURLToRealPath(url: string) {
+    return resolveRealPath(compilerHost, compilerHost.fileURLToPath(url));
+  }
+
+  function createFileSystemCache() {
+    const cache = new Map<string, CachedFile | CachedError>();
+    let changes: FileEvent[] = [];
+    return {
+      async get(path: string) {
+        for (const change of changes) {
+          const path = await fileURLToRealPath(change.uri);
+          cache.delete(path);
+        }
+        changes = [];
+        return cache.get(path);
+      },
+      set(path: string, entry: CachedFile | CachedError) {
+        cache.set(path, entry);
+      },
+      notify(changes: FileEvent[]) {
+        changes.push(...changes);
+      },
+    };
+  }
+
+  function createCompilerHost(): CompilerHost {
+    const base = host.compilerHost;
+    const overrides: Partial<CompilerHost> = {
+      readFile,
+      stat,
+      getSourceFileKind,
+    };
+    const compilerHost = Object.create(base);
+    Object.assign(compilerHost, overrides);
```

Comment:
any reason doing this vs just spread?

Reply:
It was so that the test could replace the base host's realpath implementation.

Reply:
wouldn't 
```
const compilerHost = {
   ...host.compilerHost,
 readFile,
      stat,
      getSourceFileKind,
```

achieve the same?

Reply:
Not quite `...` would copy host.compilerHost properties so any future changes to host.compilerHost would not be observed. Now normally the host.compilerHost would never change, but I did this in a test.

https://github.com/microsoft/cadl/pull/753/files#diff-84801f11cab40f293ffd92b4958894a86c9344f4bab6815ee4f648f7e099ace3R8

This is a little hacky, but I got a little fed up of plumbing options through the test infrastructure. :)

Reply:
I added a comment. If you feel strongly, I can make it so that the test passes down its realpath implementation rather than overwrites with it.

Reply:
I do feel like it be better to allow passing override in teh `createTestServerHost` instead. this feels like it will endup causing unexpected behaviors

Reply:
Yeah, you're right 😳

Fixed!